src/main/java/tech/picnic/jolo/LoaderFactoryBuilderImpl.java

```diff
@@ -64,60 +64,62 @@
   }
 
   @Override
-  public <L, R, L2 extends Record, R2 extends Record> RelationBuilder<T, L, R> oneToOne(
-      Entity<L, L2> left, Entity<R, R2> right) {
+  public <L, R, L2 extends Record, R2 extends Record, K> RelationBuilder<T, L, R, K> oneToOne(
+      Entity<L, L2, K> left, Entity<R, R2, K> right) {
     return relation(left, right)
         .oneToOne(getForeignKeySymmetric(left.getTable(), right.getTable()));
   }
 
   @Override
-  public <L, R, L2 extends Record, R2 extends Record> RelationBuilder<T, L, R> oneToZeroOrOne(
-      Entity<L, L2> left, Entity<R, R2> right) {
+  public <L, R, L2 extends Record, R2 extends Record, K> RelationBuilder<T, L, R, K> oneToZeroOrOne(
+      Entity<L, L2, K> left, Entity<R, R2, K> right) {
     return relation(left, right)
         .oneToZeroOrOne(getForeignKeySymmetric(left.getTable(), right.getTable()));
   }
 
   @Override
-  public <L, R, L2 extends Record, R2 extends Record> RelationBuilder<T, L, R> optionalOneToOne(
-      Entity<L, L2> left, Entity<R, R2> right) {
+  public <L, R, L2 extends Record, R2 extends Record, K>
+      RelationBuilder<T, L, R, K> optionalOneToOne(Entity<L, L2, K> left, Entity<R, R2, K> right) {
     return relation(left, right)
         .optionalOneToOne(getForeignKeySymmetric(left.getTable(), right.getTable()));
   }
 
   @Override
-  public <L, R, L2 extends Record, R2 extends Record> RelationBuilder<T, L, R> zeroOrOneToMany(
-      Entity<L, L2> left, Entity<R, R2> right) {
+  public <L, R, L2 extends Record, R2 extends Record, K>
+      RelationBuilder<T, L, R, K> zeroOrOneToMany(Entity<L, L2, K> left, Entity<R, R2, K> right) {
     return relation(left, right).zeroOrOneToMany(getForeignKey(right.getTable(), left.getTable()));
   }
 
   @Override
-  public <L, R, L2 extends Record, R2 extends Record> RelationBuilder<T, L, R> manyToMany(
-      Entity<L, L2> left, Entity<R, R2> right, Table<?> relation) {
-    TableField<?, Long> leftKey = getForeignKey(relation, left.getTable());
-    TableField<?, Long> rightKey = getForeignKey(relation, right.getTable());
+  public <L, R, L2 extends Record, R2 extends Record, K> RelationBuilder<T, L, R, K> manyToMany(
+      Entity<L, L2, K> left, Entity<R, R2, K> right, Table<?> relation) {
+    TableField<?, K> leftKey = getForeignKey(relation, left.getTable());
+    TableField<?, K> rightKey = getForeignKey(relation, right.getTable());
     return relation(left, right).manyToMany(leftKey, rightKey);
   }
 
-  private static <L extends Record, R extends Record> TableField<?, Long> getForeignKeySymmetric(
+  @SuppressWarnings("unchecked")
+  private static <L extends Record, R extends Record, K> TableField<?, K> getForeignKeySymmetric(
       Table<L> left, Table<R> right) {
-    TableField<?, Long> leftKey = getOptionalForeignKey(right, left).orElse(null);
-    TableField<?, Long> rightKey =
+    TableField<?, K> leftKey = (TableField<?, K>) getOptionalForeignKey(right, left).orElse(null);
```

Comment:
You can avoid casting here and elsewhere by doing 
```suggestion
    TableField<?, K> leftKey = Util.<R, L, K>getOptionalForeignKey(right, left).orElse(null);
```

Reply:
Thanks for the suggestion, I now removed the explicit castings I introduced by using this `Util` method.